src/wp-includes/blocks.php

```diff
@@ -402,37 +418,24 @@ function register_block_type_from_metadata( $file_or_folder, $args = array() ) {
 	 * instead of reading a JSON file per-block, and then decoding from JSON to PHP.
 	 * Using a static variable ensures that the metadata is only read once per request.
 	 */
-	static $core_blocks_meta;
-	if ( ! $core_blocks_meta ) {
-		$core_blocks_meta = require ABSPATH . WPINC . '/blocks/blocks-json.php';
-	}
 
 	$metadata_file = ( ! str_ends_with( $file_or_folder, 'block.json' ) ) ?
 		trailingslashit( $file_or_folder ) . 'block.json' :
 		$file_or_folder;
 
-	$is_core_block = str_starts_with( $file_or_folder, ABSPATH . WPINC );
-	// If the block is not a core block, the metadata file must exist.
-	$metadata_file_exists = $is_core_block || file_exists( $metadata_file );
-	if ( ! $metadata_file_exists && empty( $args['name'] ) ) {
-		return false;
-	}
-
-	// Try to get metadata from the static cache for core blocks.
-	$metadata = array();
-	if ( $is_core_block ) {
-		$core_block_name = str_replace( ABSPATH . WPINC . '/blocks/', '', $file_or_folder );
-		if ( ! empty( $core_blocks_meta[ $core_block_name ] ) ) {
-			$metadata = $core_blocks_meta[ $core_block_name ];
-		}
-	}
-
-	// If metadata is not found in the static cache, read it from the file.
-	if ( $metadata_file_exists && empty( $metadata ) ) {
-		$metadata = wp_json_file_decode( $metadata_file, array( 'associative' => true ) );
-	}
-
-	if ( ! is_array( $metadata ) || ( empty( $metadata['name'] ) && empty( $args['name'] ) ) ) {
+	$metadata             = null;
+	$metadata_file_exists = false;
+	$registry_metadata    = WP_Block_Metadata_Registry::get_metadata( $file_or_folder );
+
+	if ( $registry_metadata ) {
+		$metadata             = $registry_metadata;
+		$metadata_file_exists = file_exists( $metadata_file );
+	} elseif ( file_exists( $metadata_file ) ) {
+		$metadata             = wp_json_file_decode( $metadata_file, array( 'associative' => true ) );
+		$metadata_file_exists = true;
+	} elseif ( ! empty( $args['name'] ) ) {
+		$metadata = array();
+	} else {
```

Comment:
See https://github.com/WordPress/wordpress-develop/pull/7303#discussion_r1777420373. We need to keep the early return condition separate, as this scenario may theoretically also happen when reading data from files (including getting it from the registry). This is mostly about maintaining the current behavior, which is safer, because we don't know why these specific clauses were added. So let's bring back the original clause here at the end.

I think this can be fixed and simplified to:
```suggestion
	$metadata_file_exists = file_exists( $metadata_file );
	$registry_metadata    = WP_Block_Metadata_Registry::get_metadata( $file_or_folder );

	if ( $registry_metadata ) {
		$metadata = $registry_metadata;
	} elseif ( $metadata_file_exists ) {
		$metadata = wp_json_file_decode( $metadata_file, array( 'associative' => true ) );
	} else {
		$metadata = array();
	}

	if ( ! is_array( $metadata ) || ( empty( $metadata['name'] ) && empty( $args['name'] ) ) ) {
```

Reply:
Ah, right. Looking back at it, the is_array check for $metadata is essential because it's possible for wp_json_file_decode() to return false. My older code would have fataled in that case. Good catch, applied.